### PR TITLE
Systemd integration with runc, for on-demand socket activation

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
Needs rebase. Also, could you add pointers or some documentation to motivate the feature?

**Comment by shishir-a412ed** [comment]
systemd has a feature known as on-demand socket activation. Which allows systemd to only spawn services when a request for that service comes in. To give an example:

Machine 1: Fedora 23, SSH server (Not running right now)
Machine 2: Mac OS X 10.10.5

On my machine 2, I open a terminal and type `ssh IP_ADDRESS_MACHINE_1`
This request will go to machine 1 systemd, and only now systemd will spawn the SSHD service.
How systemd achieves this:

Systemd creates a socket to listen for incoming ssh requests.
When the first request comes in, systemd accepts the connection on behalf of SSHD and spawn the service. Systemd will then pass this socket to SSHD so that the future requests should directly be handled by SSHD (rather than systemd)

This patch allows systemd to have on-demand socket activation for containerized services (and not just native services).

Changes:

1) When systemd fork-exec runc, runc will get two environment variables.
     a) LISTEN_PID: the process ID which should listen on the on-demand socket.
     b) LISTEN_FDS: number of file descriptors passed. For example, LISTEN_FDS=2 would mean systemd would pass 3,4 {in addition to 0,1,2} to support socket activation.
     c) runc by default will inherit all open FDS including the ones passed for socket activation.

2) Runc would set the environment for the container: 
    a) LISTEN_FDS="the value set by systemd".
    b) LISTEN_PID=1
        LISTEN_PID would be set to 1 as container init process would be the one inheriting these FDS.

3) runc will pass the required FDS to container init process based on the value of LISTEN_FDS.

File: start.go [hunk 1]
```diff
@@ -100,6 +123,13 @@ func setupSdNotify(spec *specs.LinuxSpec, notifySocket string) {
 	spec.Process.Env = append(spec.Process.Env, fmt.Sprintf("NOTIFY_SOCKET=%s", notifySocket))
 }
 
+// If systemd is supporting on-demand socket activation, this function will add support
+// for on-demand socket activation for the containerized service.
+func setupSocketActivation(spec *specs.LinuxSpec, listenFds string) {
+	spec.Process.Env = append(spec.Process.Env, fmt.Sprintf("LISTEN_FDS=%s", listenFds))
```
Comment (crosbymichael) [inline start.go]: This and the line below can be done in one line:

```go
spec.Process.Env = append(spec.Process.Env, 
    fmt.Sprintf("LISTEN_FDS=%s", listenFds),  
    "LISTEN_PID=1")
```

Comment (shishir-a412ed) [inline start.go]: @crosbymichael Fixed.

**Comment by crosbymichael** [comment]
LGTM

ping @vbatts @mrunalp

File: start.go [hunk 2]
```diff
@@ -79,6 +89,19 @@ func startContainer(context *cli.Context, spec *specs.LinuxSpec) (int, error) {
 	// that created it.
 	defer destroy(container)
 	process := newProcess(spec.Process)
+
+	//support for on-demand socket activation: pass file descriptors into the container.
```
Comment (mrunalp) [inline start.go]: // Support on-demand socket activation by passing file descriptors...

Comment (shishir-a412ed) [inline start.go]: @mrunalp Fixed.

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**